File: neet_code_150/src/greedy/max_subarray.rs
// 最小部分列

// 初見では解けなかった
struct Solution {}
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        3
    }
}

struct SolutionAns {}
impl SolutionAns {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut res = nums[0];
        let mut sum = 0;

        for n in nums {
            if sum < 0 {
                sum = 0;
            }

            sum += n;
            res = res.max(sum);
        }

        res
    }
}

fn main() {
    let case_1 = vec![-2, 1, -3, 4, -1, 2, 1, -5, 4];
    let case_2 = vec![1];
    let case_3 = vec![5, 4, -1, 7, 8];

    println!("case_1: {}", SolutionAns::max_sub_array(case_1.clone()));
    println!("case_2: {}", SolutionAns::max_sub_array(case_2.clone()));
    println!("case_3: {}", SolutionAns::max_sub_array(case_3.clone()));
}

File: neet_code_150_cpp/greedy/max_sub_array.cpp
#include <iostream>
#include <vector>

class SolutionAns {
public:
    //Kadaneのアルゴリズム
    int maxSubArray(std::vector<int>& nums) {
        int curr = nums[0];
        int result = nums[0];
        
        for (int i = 1; i < nums.size(); i++) {
            //currをmax(curr + nums[i],nums[i])で更新する
            curr = std::max(curr + nums[i], nums[i]);
            // resultをmax(result, curr)で更新する
            result = std::max(result, curr);
        }
        
        return result;
    }
};

int main(void) {
    auto case_1 = std::vector{-2, 1, -3, 4, -1, 2, 1, -5, 4};
    auto case_2 = std::vector{1};
    auto case_3 = std::vector{5, 4, -1, 7, 8};

    SolutionAns s_ans;
    std::cout << s_ans.maxSubArray(case_1) << std::endl;
    std::cout << s_ans.maxSubArray(case_2) << std::endl;
    std::cout << s_ans.maxSubArray(case_3) << std::endl;
}
